File: sdk/go/typed.go
package anyvali

import "fmt"

// TypedParseResult is a generic parse result with typed data.
type TypedParseResult[T any] struct {
	Success bool
	Data    T
	Issues  []ValidationIssue
}

// TypedParse parses input using the given schema and returns a typed result.
// Returns an error if validation fails or the result cannot be cast to T.
func TypedParse[T any](s Schema, input any) (T, error) {
	result, err := s.Parse(input)
	if err != nil {
		var zero T
		return zero, err
	}
	typed, ok := result.(T)
	if !ok {
		var zero T
		return zero, fmt.Errorf("anyvali: type assertion failed: expected %T, got %T", zero, result)
	}
	return typed, nil
}

// TypedSafeParse parses input and returns a TypedParseResult with typed data.
func TypedSafeParse[T any](s Schema, input any) TypedParseResult[T] {
	result := s.SafeParse(input)
	if !result.Success {
		return TypedParseResult[T]{Success: false, Issues: result.Issues}
	}
	typed, ok := result.Data.(T)
	if !ok {
		var zero T
		return TypedParseResult[T]{Success: false, Data: zero, Issues: []ValidationIssue{{
			Code:    "type_assertion_failed",
			Message: fmt.Sprintf("Cannot convert %T to target type", result.Data),
			Path:    []any{},
		}}}
	}
	return TypedParseResult[T]{Success: true, Data: typed}
}

File: sdk/rust/src/typed.rs
use serde::de::DeserializeOwned;
use serde_json::Value;

use crate::schema::Schema;
use crate::schemas::{
    AnySchema, ArraySchema, BoolSchema, IntSchema, NullSchema, NumberSchema, StringSchema,
    UnknownSchema,
};
use crate::types::{ValidationError, ValidationIssue};

/// Trait for schemas that know their output Rust type.
/// Provides a typed parse method that deserializes the validated result.
pub trait TypedSchema: Schema {
    type Output: DeserializeOwned;

    /// Parse and validate the input, then deserialize the result to the associated Output type.
    fn parse_typed(&self, input: &Value) -> Result<Self::Output, ValidationError> {
        let value = self.parse(input)?;
        serde_json::from_value(value).map_err(|e| ValidationError {
            issues: vec![ValidationIssue {
                code: "type_conversion".to_string(),
                path: vec![],
                expected: std::any::type_name::<Self::Output>().to_string(),
                received: format!("{}", e),
                meta: None,
            }],
        })
    }
}

/// Parse input using any schema and deserialize the result to a specific type.
///
/// This is useful when working with `dyn Schema` trait objects where the
/// concrete type (and thus `TypedSchema`) is not known at compile time.
pub fn parse_as<T: DeserializeOwned>(
    schema: &dyn Schema,
    input: &Value,
) -> Result<T, ValidationError> {
    let value = schema.parse(input)?;
    serde_json::from_value(value).map_err(|e| ValidationError {
        issues: vec![ValidationIssue {
            code: "type_conversion".to_string(),
            path: vec![],
            expected: std::any::type_name::<T>().to_string(),
            received: format!("{}", e),
            meta: None,
        }],
    })
}

// --- TypedSchema implementations for concrete schema types ---

impl TypedSchema for StringSchema {
    type Output = String;
}

impl TypedSchema for BoolSchema {
    type Output = bool;
}

impl TypedSchema for NumberSchema {
    type Output = f64;
}

impl TypedSchema for IntSchema {
    type Output = i64;
}

impl TypedSchema for NullSchema {
    /// Null parses to `()` since `serde_json::Value::Null` deserializes to `()`.
    type Output = ();
}

impl TypedSchema for AnySchema {
    type Output = Value;
}

impl TypedSchema for UnknownSchema {
    type Output = Value;
}

impl TypedSchema for ArraySchema {
    type Output = Vec<Value>;
}
